docs/plutus-report/scripting.tex

```diff
@@ -111,7 +115,127 @@ \subsection{Recursive values}
+
+\paragraph{Passing arguments to builtin functions}
+
+Builtin functions must of course be provided with a type signature (since we have a typed language), and a way to evaluate them.
+The natural way to provide such evaluators is as functions --- but these functions must be able to operate on \gls{plutus-core} terms!
+For example, an evaluator for ``plus'' must be able to take two \emph{terms} of type \code{Integer} and turn them into another term.
+But a simple implementation of ``plus'' just talks about \emph{integers}, not terms.
+
+Solving this problem in generality is difficult, so we make a key simplification: builtin functions may only accept arguments of builtin types (with one exception, as we will see shortly).
+Along with the fact that (since \gls{plutus-core} is eager) function arguments are evaluated to values before being passed to the function, this means that the arguments to builtin evaluators will always be \emph{constants}, which are easy to unwrap into their underlying values.
+
+\paragraph{Polymorphic builtin functions}
+\label{sec:polymorphic-builtins}
+
+The exception to the above rule about builtin function types is for \emph{polymorphic} builtin functions.
+These must necessarily be able to process arbitrary terms (since they can be instantiated at any type, not just a builtin type).
+For example, \code{ifThenElse} must be able to produce a value of type \code{Integer -> Integer}, even though this is not a builtin type.
+
+It is possible in principle to use polymorphic \emph{evaluators} to implement polymorphic builtin functions \autocite{lindley2012embedding}.
+Unfortunately this is quite a heavyweight and complex encoding for such a key usecase.
+
+Instead, we require evaluators for polymorphic functions to be \emph{parametric}, which means in particular that they do not inspect their polymorphic arguments.
+We can thus pass the \emph{term itself} as the polymorphic argument, instead of needing to extract a value of an appropriate type in the metalanguage, as we do for other arguments.
+Thus, an evaluator for \code{const :: forall a b . a -> b -> a} takes two \emph{terms} and returns the first one.
+
+\paragraph{Saturated builtin application}
+
+When including builtin function application in the AST, we have two options:
+\begin{enumerate}
+\item
+  Builtin functions can stand alone as a value with the builtin's function type, and are applied to their arguments with normal function application.
+\item
+  Builtin functions must always appear \emph{saturated}, that is, with all of their arguments.
+\end{enumerate}
+
+There choice is fairly minor, but has a few consequences:
+\begin{enumerate}
+\item
+  Unsaturated builtins can be partially applied, saturated builtins cannot be --- but we can easily work around this by wrapping the builtin in a lambda, so this does not matter in practice.
+\item
+  Saturated builtins are easier to evaluate.
```

Comment:
"easier to evaluate" is somewhat ambiguous. Maybe something like "are easier to implement an efficient evaluator for" or the like?